documentation/diagnosing-slow-builds.md

```diff
@@ -0,0 +1,44 @@
+# Diagnosing slow builds
+
+No build is ever fast enough. It's often worth investing some effort on understanding where MSBuild spends its time when building your code, and thinking about whether it can be improved.
+
+[Obligatory XKCD](https://www.xkcd.com/303/)
+
+Logs at MSBuild’s “normal” level don't capture most of the really interesting information, though.
+
+Try [capturing a binary log](https://aka.ms/msbuild/binlog) in your build; that preserves detailed information and allows for offline analysis. Changing log verbosity levels can affect build performance, but doesn't usually cause dramatic slowdown for full builds. Cross-check the results of with-logging and without-logging builds to make sure they're not completely out of whack before delving deeply into analyzing the binary log.
+
+## Setting expectations
+
+Do not expect to see 100% processor utilization during the entire build. Builds are often constrained by disk I/O, and the build process is also constrained by the dependencies between projects that are being built. If 100 projects depend on a common library, that library must be built before they can start building.
```

Comment:
Is there anything out there that prints the critical path, so people can easily see it and work on reducing it? I regularly print out the p2p graph with [this thing](https://github.com/microsoft/MSBuildStaticGraphTester/tree/master/src/GraphGen) whenever I need to debug static graph, but I guess it could be used to visually look and identity critical paths. It can also be reasonably easily changed to add an option to color the critical path. Or to prune everything else out.

Reply:
"Is there anything out there that prints the critical path"

I too am interested in this, recently we spun up a 64 Core VB with 128GB of RAM in Azure, we allocated a 64GB RAM Drive to try and rule out any I/O limitations. Historically I have used a utility that I open sourced awhile back (https://github.com/aolszowka/MsBuildProjectReferenceDependencyGraph) however based on my graphs I don't think this is the issue we are encountering.